python/mxnet/optimizer.py

```diff
@@ -670,21 +672,37 @@ def __init__(self, eps=1e-7, **kwargs):
         self.float_stable_eps = eps
 
     def create_state(self, index, weight):
-        return zeros(weight.shape, weight.context)  # history
+        return zeros(weight.shape, weight.context, stype=weight.stype)  # history
 
     def update(self, index, weight, grad, state):
         assert(isinstance(weight, NDArray))
         assert(isinstance(grad, NDArray))
         self._update_count(index)
         lr = self._get_lr(index)
         wd = self._get_wd(index)
-
+        save_grad_stype = grad.stype
         grad = grad * self.rescale_grad
         if self.clip_gradient is not None:
             grad = clip(grad, -self.clip_gradient, self.clip_gradient)
         history = state
-        history[:] += (grad * grad)
-        weight[:] += -lr * (grad / sqrt(history + self.float_stable_eps) + wd * weight)
+        save_history_stype = history.stype
+
+        is_sparse = True if weight.stype != 'default' or grad.stype != 'default' else False
+
+        if is_sparse:
+            history[:] = op.elemwise_add(history, op.square(grad))
+            assert history.stype == save_history_stype
+            srt = op.sqrt(_internal._scatter_plus_scalar(history, self.float_stable_eps))
+            assert srt.stype == save_history_stype
+            div = _internal._scatter_elemwise_div(grad, srt)
+            assert div.stype == grad.stype
+        else:
+            history[:] += square(grad)
+            div = grad / sqrt(history + self.float_stable_eps)
+
+        weight[:] += (div + weight * wd) * -lr
```

Comment:
Instead of `weight * wd`, it should be `sparse.retain(weight, grad.indices) * wd` since we're only updating the row slices appeared in grad.indices. Otherwise, the update is not sparse - after one epoch each update touches a million rows if you use `weight` directly.

Reply:
Ah, that version made sense! thanks!